targets/TARGET_NORDIC/TARGET_NRF5/TARGET_SDK11/drivers_nrf/rng/nrf_drv_rng.c

```diff
@@ -0,0 +1,250 @@
+/* Copyright (c) 2015 Nordic Semiconductor. All Rights Reserved.
+ *
+ * The information contained herein is property of Nordic Semiconductor ASA.
+ * Terms and conditions of usage are described in detail in NORDIC
+ * SEMICONDUCTOR STANDARD SOFTWARE LICENSE AGREEMENT.
+ *
+ * Licensees are granted free, non-transferable use of the information. NO
+ * WARRANTY of ANY KIND is provided. This heading must NOT be removed from
+ * the file.
+ *
+ */
+
+#include <stdint.h>
+#include <stddef.h>
+
+#include "nrf_drv_rng.h"
+#include "nrf_assert.h"
+#include "nrf_drv_common.h"
+#include "nordic_common.h"
+#include "nrf_error.h"
+#include "nrf_assert.h"
+#ifdef SOFTDEVICE_PRESENT
+#include "nrf_sdm.h"
+#include "nrf_soc.h"
+#else
+#include "app_fifo.h"
+#include "app_util_platform.h"
+
+static __INLINE uint32_t fifo_length(app_fifo_t * p_fifo)
+{
+    uint32_t tmp = p_fifo->read_pos;
+    return p_fifo->write_pos - tmp;
+}
+
+#define FIFO_LENGTH(fifo) fifo_length(&(fifo))  /**< Macro for calculating the FIFO length. */
+
+#endif // SOFTDEVICE_PRESENT
+typedef struct
+{
+    nrf_drv_state_t state;
+#ifndef SOFTDEVICE_PRESENT
+    app_fifo_t rand_pool;
+    uint8_t    buffer[RNG_CONFIG_POOL_SIZE];
+#endif // SOFTDEVICE_PRESENT
+} nrf_drv_rng_cb_t;
+
+static nrf_drv_rng_cb_t m_rng_cb;
+#ifndef SOFTDEVICE_PRESENT
+static const nrf_drv_rng_config_t m_default_config = NRF_DRV_RNG_DEFAULT_CONFIG;
+static void rng_start(void)
+{
+    if (FIFO_LENGTH(m_rng_cb.rand_pool) <= m_rng_cb.rand_pool.buf_size_mask)
+    {
+        nrf_rng_event_clear(NRF_RNG_EVENT_VALRDY);
+        nrf_rng_int_enable(NRF_RNG_INT_VALRDY_MASK);
+        nrf_rng_task_trigger(NRF_RNG_TASK_START);
+    }
+}
+
+
+static void rng_stop(void)
+{
+    nrf_rng_int_disable(NRF_RNG_INT_VALRDY_MASK);
+    nrf_rng_task_trigger(NRF_RNG_TASK_STOP);
+}
+
+
+#endif // SOFTDEVICE_PRESENT
+
+
+ret_code_t nrf_drv_rng_init(nrf_drv_rng_config_t const * p_config)
+{
+    uint32_t result;
+
+    if (m_rng_cb.state == NRF_DRV_STATE_UNINITIALIZED)
+    {
+#ifndef SOFTDEVICE_PRESENT
+
+        result = app_fifo_init(&m_rng_cb.rand_pool, m_rng_cb.buffer, RNG_CONFIG_POOL_SIZE);
+
+        if (p_config == NULL)
+        {
+            p_config = &m_default_config;
+        }
+
+        if (result == NRF_SUCCESS)
+        {
+            if (p_config->error_correction)
+            {
+                nrf_rng_error_correction_enable();
+            }
+
+            nrf_drv_common_irq_enable(RNG_IRQn, p_config->interrupt_priority);
+
+            nrf_rng_shorts_disable(NRF_RNG_SHORT_VALRDY_STOP_MASK);
+
+            rng_start();
+            m_rng_cb.state = NRF_DRV_STATE_INITIALIZED;
+        }
+#else
+        UNUSED_VARIABLE(p_config);
+        uint8_t softdevice_is_enabled;
+        result = sd_softdevice_is_enabled(&softdevice_is_enabled);
+
+        if (softdevice_is_enabled)
```

Comment:
It seems to be safer to only take this branch if `sd_softdevice_is_enabled` has been successful _and_ `softdevice_is_enabled` is set.

Reply:
Note this is their SDK thus should be sent upstream if needed